dbus/src/message/parser.rs

```diff
@@ -0,0 +1,290 @@
+            Error::TooLong => {
+                write!(f, "MatchRules may not contain more than {} elements", MAX_RULE_TOKENS)
+            }
+            Error::UnknownType => {
+                write!(f, "Unsupported message type")
+            }
+            Error::UnknownKey => {
+                write!(f, "Unknown key used")
+            }
+            Error::BadBoolean => {
+                write!(f, "Got bad boolean value")
+            }
+            Error::BadConversion(err) => {
+                write!(f, "Error while converting: {}", err)
+            }
+        }
+    }
+}
+
+impl stdError for Error {}
+
+impl<'a> TryFrom<&'a str> for MessageType {
+    type Error = Error;
+
+    fn try_from(value: &'a str) -> Result<Self, <crate::message::MessageType as TryFrom<&'a str>>::Error> {
+        match value {
+            "error" => Ok(MessageType::Error),
+            "method_call" => Ok(MessageType::MethodCall),
+            "method_return" => Ok(MessageType::MethodReturn),
+            "signal" => Ok(MessageType::Signal),
+            _ => Err(Error::UnknownType)
+        }
+    }
+}
+
+/// Key-Value-pair
+pub type TokenRule<'a> = (&'a str, &'a str);
+/// Fixed size buffer for match rule tokens
+pub type TokenBuffer<'a> = ArrayVec<TokenRule<'a>, MAX_RULE_TOKENS>;
+
+#[derive(Clone, Debug)]
+/// Tokenizes a match rule into key-value-pairs
+struct Tokenizer<'a> {
+    text: &'a str,
+}
+
+impl<'a> Tokenizer<'a> {
+    /// Builds a new tokenizer for the input &str
+    pub fn new(text: &'a str) -> Self {
+        Self {
+            text,
+        }
+    }
+
+    /// Parse the key part of the key-value-pair. This is rather easy as all keys are rather
+    /// easily defined and may only contain `[a-z0-9]` so we can simply split at `'='`.
+    fn key(&self) -> (&'a str, &'a str) {
+        let index = self.text.find('=').unwrap_or_else(|| self.text.len());
+
+        ((&self.text[..index]).trim(), &self.text[index + 1..])
+    }
+
+    /// Parses values as generic strings.
+    /// This does not do any validation (yet) with regards to supported characters.
+    fn value(&self) -> (&'a str, &'a str) {
+        let mut i = 0;
+        let mut quoted = false;
+        let mut escape = false;
+
+        for c in self.text.chars() {
+            match c {
+                '\'' if !escape => {
+                    quoted = !quoted;
+                }
+                ',' if !quoted => {
+                    break;
+                }
+                '\\' if !quoted => {
+                    escape = true;
+                    i += 1;
+                    continue;
+                }
+                _ => {}
+            }
+            escape = false;
+
+            i += 1;
+        }
+
+        // Skip comma if there is still space in the buffer
+        if self.text.len() == i {
+            ((&self.text[..i]).trim(), &self.text[i..])
+        } else {
+            ((&self.text[..i]).trim(), &self.text[i + 1..])
+        }
```

Comment:
Less copy-paste with

    // Skip comma if there is still space in the buffer
    let j = if self.text.len() == i { i } else { i + 1 };
    (&self.text[..i]).trim(), &self.text[j..])